csrc/scheduler/mma_utils.h

```diff
@@ -242,34 +244,41 @@ class DataWrapperOpt {
   }
 };
 
-// This struct hold properties of a Mul and Sum pair
-// which can possibly be replaced a Mma op. This struct
-// can be be created (partially) from a Mma op.
-struct MulSumProperties {
-  // The Mul amd Sum op which can be replaced by a Mma op.
-  struct MulAndSumOps {
-    BinaryOp* mop = nullptr;
-    ReductionOp* redop = nullptr;
-  };
-
-  // The inputs/ouputs to the possible Mma Op or the actual Mma op.
-  struct InputsOutputs {
-    TensorView* a = nullptr;
-    TensorView* b = nullptr;
-    TensorView* out = nullptr;
-  };
-
-  // The broadcasts which feed the Mma op/Mul-Sum pair.
-  struct Broadcasts {
-    BroadcastOp* bcast_a = nullptr;
-    BroadcastOp* bcast_b = nullptr;
-  };
-
-  MulAndSumOps mulsumops;
-  InputsOutputs insouts;
-  Broadcasts bcasts;
+//! This represents a single matmul operation, without a prologue or epilogue.
+//! Each matmul has two inputs which might not be fusion inputs: A and B. It
+//! also has one output, which can be Float or reduced precision. For MatmulOp
+//! and LinearOp, the output is the same dtype as the inputs; so output does not
+//! necessarily correspond to the output of a translated MmaOp and it might not
+//! be a fusion output.
+struct MatmulPattern {
+  TensorView* A;
+  TensorView* B;
+  // This is not necessarily a Fusion output, but rather is the immediate output
+  // representing a matmul in the current Fusion. The definition of this tensor
+  // determines what kind of translation is needed, if any. Possible definition
+  // Expr types are: MmaOp, ReductionOp (for mul-sum patterns), MatmulOp, and
+  // LinearOp.
+  TensorView* output;
+
+  //! If the pattern is not already represented by an MmaOp, for example if
+  //! there is a MatmulOp instead, this function modifies the fusion to insert
+  //! an MmaOp. TensorViews A and B are unchanged, but this->output might be
+  //! updated to reflect the replacement tensor.
```

Comment:
> but this->output might be updated to reflect the replacement tensor.

Is it? I don't see it in the definition of `translateToMmaOp`.

Reply:
This would be the case for `MatmulOp` which ends with a downcast. If all uses of `output` are upcasts back to `Float` then we can remove the casts. In that case we would update output to point to the Float output since the `Half` output is no longer alive in the fusion.